torch/csrc/distributed/rpc/tensorpipe_utils.h

```diff
@@ -23,6 +23,54 @@ const c10::Stream& getStreamForDevice(
     const std::vector<c10::Stream>& streams,
     const c10::Device& device);
 
+// Inspired by c10/core/impl/DeviceGuardImplInterface.h.
+
+class TensorpipeDeviceTypeConverter {
+ public:
+  // Ideally we'd want this to also return a tensorpipe::Message::Tensor object
+  // but we cannot forward-declare that class (because it's nested), and we
```

Comment:
It's indeed a little weird to add tensors to message and then use `TORCH_INTERNAL_ASSERT` to make sure that the tensor is indeed inserted. 

Do we expect other files to access this `TensorpipeDeviceTypeConverter`? If not, maybe it can be completed implemented in the `.cpp` file and directly using `tensorpipe::Message::Tensor` there?

Reply:
I should have made it clearer but, yes, in the next PR (#59377) I am moving the CUDA subclass of this to a separate file, thus the abstract base class must be defined in a header.